### PR TITLE
OAPA: self-calibration, auto-scaled corrections, and convergence hardening

### PR DESCRIPTION
**Pull request by michelebergo**
## Summary

This PR hardens the OAPA automated polar alignment support with a self-calibration routine, auto-scaled corrections, and several convergence guards. Everything has been validated over multiple field-test iterations with beta testers on real hardware (details below).

Rebased on 2.2.6.3; version bumped to 2.2.7.0 with a full changelog entry. Avalon UPAS behavior is unchanged, and all existing settings keys are preserved.

### OAPA Self-Calibration (new panel in the OAPA control dock)
- Per axis: priming move (+45', absorbs pending backlash) → solve → forward leg (+45') → solve → reversal leg (−45') → solve → reverse leg (−45') → solve. Net commanded motion is zero.
- The two backlash-free legs yield the **calibration factor**; the reversal leg comes up short by exactly the mechanical backlash, which is **measured automatically** and persisted into the backlash compensation settings on Apply.
- **cos(field altitude) correction on azimuth displacements** — a base rotation of θ moves a field at altitude h by θ·cos(h). Without this, the discovered factor depended on where the scope pointed (field data: same hardware yielded factor 52 at alt ~79° vs 16 near the pole). The azimuth leg refuses to run when cos(alt) < 0.25 and directs the user toward the pole.
- Baseline solve before any motion (unsolvable field aborts at zero cost; mid-sequence failures drive the accumulated motion back), forward/reverse leg asymmetry warning (>20%), automatic Reverse Az/Alt flag correction with a single retry.

### Automated adjustment improvements
- **Per-cycle correction limit auto-scales with the measured error** (80% of the current total error, floor 5', ceiling from a new "Max correction per cycle" setting, default 30'). Multi-degree initial errors converge in a handful of cycles with zero configuration; the final approach stays gentle.
- **Identification probes scale with the measured error** (15%, clamped between 1 and half the cap) so probes don't drown in solve noise on large errors, and a 75% correction candidate is evaluated alongside the existing ones.
- **Backlash clearing is skipped when the commanded nudge is smaller than the compensation** — field logs showed the ±compensation out-and-back excursion injecting error oscillation during the fine approach (5'+ compensation vs sub-arcminute nudges). Manual absolute moves still always clear.

### Convergence guards
- **Runaway guard**: if corrections make the total error worse for 3 consecutive stable measurements, motor movement halts and the alignment pauses with an error notification (wrong calibration/backlash no longer chases the error away from the pole). Resume keeps the error display active for manual adjustment.
- **Auto-finish requires 2 consecutive solves below tolerance** (a single lucky solve can no longer end a non-converged procedure); motors hold still between confirmation solves.

### Fixes and quality of life
- **Alignment Tolerance accepts decimals** — the `UpdateSourceTrigger=PropertyChanged` binding re-parsed on every keystroke and swallowed the decimal separator; now commits on focus loss with `StringFormat 0.##` (instruction template + Options).
- OAPA firmware version detection: the reference firmware reports its version in the status frame (optional `|V:x.y.z|` field, backward compatible); the plugin logs it and warns once when the firmware is missing or older than the recommended version.
- OAPA Home Position panel (Set Home / Go Home), last-known-good COM port tried first on reconnect, ratio-aware motion completion timeouts, separate Y-axis backlash compensation, "Gear Ratio" renamed to "Calibration Factor" in the UI (settings keys unchanged).

## Firmware reference

The OAPA reference firmware (ESP32/GRBL-flavoured protocol) now has an official public home: https://github.com/michelebergo/oapa-firmware — linked from the OAPA control panel, the options tooltip, and the FAQ. The serial protocol is documented there.

## Field validation

Multiple test iterations with beta testers on real OAPA hardware (ESP32 + TMC2209, high gear-reduction platform):

- From **1°14' initial error to 0.36'** in ~6 minutes / 19 solves, fully hands-off
- From **8' to 0.45'** in ~3 minutes
- Best runs reached **0.27' (16")** at a 0.35' tolerance target
- Calibration reproducibility after the cos(alt) fix: two independent calibrations agreed within ~4% (factors 14.91 vs 14.31, backlash 5.05' vs 5.12') — before the fix the same hardware produced 52.34 vs 16.46 depending on pointing
- The runaway guard fired correctly in a real (non-simulated) fault case and stopped the excursion

## Testing

- 51/51 NUnit tests pass, including a new `AutomatedAdjustmentControllerTest` suite covering cap clamping, error-scaled probes (far from and near the pole), and a closed-loop convergence test verifying that a raised cap reduces iterations on a 1.8° error
- `InternalsVisibleTo` added to the plugin project for the new controller tests

Happy to split this into smaller PRs if you prefer, and to adjust anything to your conventions.

### CODE REVIEW

**Comment by isbeorn** [comment]
This has grown well beyond an OAPA extension. It combines at least five independently releasable changes:

1. OAPA self-calibration and home support.
2. Shared automated-controller tuning and runaway detection.
3. Shared serial connection, firmware, and movement-completion changes.
4. Global auto-finish behavior changes.
5. Decimal alignment-tolerance UI fixes.

That blast radius makes regressions difficult to attribute and forces reviewers to validate manual alignment, UPAS/Avalon, serial discovery, sequence execution, plugin composition, and OAPA hardware in one change.

Please split this first. The OAPA PR should principally remain under PolarAlignment/OAPA, with only the smallest capability boundary needed elsewhere. Shared behavior changes should be separate PRs with dedicated regression tests.

File: PolarAlignment/Instructions/PolarAlignment.cs [hunk 1]
```diff
@@ -622,7 +642,32 @@ await messageBroker.Publish(
                                     sw.Reset();
                                 }
                                 localCTS.Token.ThrowIfCancellationRequested();
-                                await TPAPAVM.MoveCloser(progress, localCTS.Token);
+
+                                // Runaway detection: compare against the previous stable measurement.
+                                if (previousTotalErrorMinutes.HasValue && totalErrorMinutes > previousTotalErrorMinutes.Value + 0.05) {
```
Comment (isbeorn) [inline PolarAlignment/Instructions/PolarAlignment.cs:647]: This is the shared runtime path for sequence and imaging-tab execution, yet the new branch diagnoses OAPA calibration and pauses the entire alignment.

It also runs when no automated system is active, when automated adjustment is disabled, or when no move occurred. Consequently, solve noise or manual adjustment can pause a manual alignment and incorrectly blame OAPA calibration.

If generic runaway protection is wanted, implement it inside AutomatedAdjustmentController, evaluate only observations following an executed plan, and submit it separately with manual, UPAS, and OAPA regression tests.

File: PolarAlignment/TPAPAVM.cs [hunk 2]
```diff
@@ -298,6 +298,14 @@ public async Task MoveCloser(IProgress<ApplicationStatus> progress, Cancellation
                 return;
             }
 
+            // Per-cycle move cap auto-scales with the current measured error so multi-degree
+            // initial errors converge in a handful of cycles without user intervention, while the
+            // final approach stays gentle. The "Max correction per cycle" setting acts as a
+            // safety ceiling on top of the auto-scaled value.
+            var currentTotalErrorArcmin = Math.Abs(PolarErrorDetermination.CurrentMountAxisTotalError.ArcMinutes);
```
Comment (isbeorn) [inline PolarAlignment/TPAPAVM.cs:305]: MoveCloser is shared by every alignment system. Auto-scaling the cap here means an OAPA convergence change also makes UPAS issue materially different commands, despite the PR providing no UPAS-specific evidence or tests.
Keep OAPA tuning in an OAPA movement policy, or make correction limits an explicit capability supplied by the selected system. A generic controller-policy change belongs in a separate PR and needs coverge for every consumer.

File: PolarAlignment/UniversalPolarAlignmentBase.cs [hunk 3]
```diff
@@ -15,6 +17,26 @@ public abstract partial class UniversalPolarAlignmentBase : IPolarAlignmentSyste
         protected virtual int ScanReadTimeout => 1000;
         protected virtual int ScanWriteTimeout => 1000;
         protected virtual bool ClearBufferOnConnect => false;
+        // Some boards (ESP32 on CH340/CP2102) auto-reset when the host opens the port and need
+        // ~1–2 s before they can answer the status query. Override this to give the firmware
+        // time to boot and emit its banner before the connection probe is sent.
+        protected virtual int PostOpenDelayMs => 100;
```
Comment (isbeorn) [inline PolarAlignment/UniversalPolarAlignmentBase.cs:23]: This OAPA extension adds boot delays, retry counts, preferred-port persistence, firmware-version parsing, firmware URLs, UI notifications, target rounding, completion tolerances, and stuck tolerances to the shared transport base.
These are several distinct device policies, not universal serial transport concepts. The defaults may preserve Avalon behavior today, but the base class has become substantially harder to understand and now imports UI notification behavior.
Keep the transport primitive small. Put OAPA connection probing, firmware compatibility, and integer-step movement semantics in OAPA-specific collaborators or overrides without expanding the shared contract this broadly. Serial-framework changes should be reviewed separately.

File: PolarAlignment/Avalon/UniversalPolarAlignmentVM.cs [hunk 4]
```diff
@@ -96,5 +96,10 @@ public override float XBacklashCompensation {
                 RaisePropertyChanged();
             }
         }
+
+        public override float YBacklashCompensation {
```
Comment (isbeorn) [inline PolarAlignment/Avalon/UniversalPolarAlignmentVM.cs:100]: Adding an abstract Y-backlash property forces Avalon to implement a silent getter returning zero and a no-op setter. That is a strong signal that an OAPA capability has been promoted into the wrong shared abstraction.

Keep Y backlash on the OAPA VM, or model optional axis-compensation capability explicitly. Do not make unrelated implementations pretend to support a feature they do not have.

File: NINA.Plugins.PolarAlignment.Test/PolarErrorDeterminationTest.cs [hunk 5]
```diff
@@ -35,9 +35,9 @@ public void PolarErrorDetermination_InitialMountError_SomeDeclinationError_Error
             var position3 = new Position(solve3, 0, latitude, longitude, elevation, refraction);
             
 
-            var error = new PolarErrorDetermination(new PlateSolving.PlateSolveResult() {  Coordinates = solve3 }, position1, position2, position3, latitude, longitude, elevation, refraction, true);
+            var error = new PolarErrorDetermination(new PlateSolving.PlateSolveResult() {  Coordinates = solve3 }, position1, position2, position3, latitude, longitude, elevation, refraction, true, 0);
```
Comment (isbeorn) [inline NINA.Plugins.PolarAlignment.Test/PolarErrorDeterminationTest.cs:38]: The constructor already defaults declinationSpreadArcsec to zero, so explicitly adding 0 across this test file changes no behavior and tests none of the OAPA work.

Please revert this noise. In a PR with this much surface area, unrelated mechanical churn makes it materially harder to identify meaningful test coverage.

File: PolarAlignment/OAPA/UniversalPolarAlignmentOAPAVM.cs [hunk 6]
```diff
@@ -1,10 +1,70 @@
...
+using NINA.PlateSolving.Interfaces;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Windows;
 
 namespace NINA.Plugins.PolarAlignment.OAPA {
     public partial class UniversalPolarAlignmentOAPAVM : UniversalPolarAlignmentBaseVM {
```
Comment (isbeorn) [inline PolarAlignment/OAPA/UniversalPolarAlignmentOAPAVM.cs:23]: This file grows from 148 to 629 lines and now owns settings, WPF state, hardware movement, image capture, plate solving, retry policy, calibration geometry, direction inference, recovery, and result application.

Extract an OapaCalibrationService with injected motion and capture/solve boundaries, plus pure calibration geometry returning a typed result. The VM should only expose commands and observable state. This dramatically reduces both the PR’s conceptual blast radius and the amount of hardware behavior that can only be tested through WPF.

File: PolarAlignment/OAPA/UniversalPolarAlignmentOAPAVM.cs [hunk 7]
```diff
@@ -144,5 +213,417 @@ public int YHoldPercent {
...
             }
         }
+
+        // ----- Home position (plugin-side, firmware-agnostic) -----
+        // Stored in plugin settings rather than firmware so it works with any OAPA-compatible
+        // controller. Positions are in the same axis units shown in the control panel; note the
+        // controller's counter restarts at 0 on power-up, so the stored home is valid for the
+        // current power session (or as long as the platform is not moved while unpowered).
+
+        public float HomeX {
```
Comment (isbeorn) [inline PolarAlignment/OAPA/UniversalPolarAlignmentOAPAVM.cs:223]: The controller counters reset on power-up, but these absolute coordinates remain persisted and actionable. After a controller reset, Go Home can move to coordinates from a previous session.

Keep home state connection/session-scoped unless the firmware exposes a persistent origin or session identifier. Stale home coordinates must be invalidated rather than merely documented.

**Review by isbeorn — CHANGES_REQUESTED** [review]

**Comment by michelebergo** [comment]
Thanks for the thorough review — all points taken, and they're fair. Agreed that the blast radius grew far beyond an OAPA extension.

I'll close this PR and split the work into independently reviewable pieces, roughly in this order:

1. **Decimal alignment-tolerance UI fix** — smallest, standalone.
2. **Auto-finish confirmation (2 consecutive solves below tolerance)** — with a dedicated test.
3. **OAPA self-calibration + home support** — scoped under `PolarAlignment/OAPA` as requested. This will include the structural feedback: calibration logic extracted into an `OapaCalibrationService` with injected motion and capture/solve boundaries plus pure geometry returning a typed result (VM reduced to commands + observable state); home state becomes session-scoped and invalidated on reconnect instead of persisted across power cycles; Y-backlash stays on the OAPA VM; and the connection-probing / firmware-compatibility / integer-step policies move into OAPA-specific overrides so the shared serial base stays a small transport primitive.
4. **Shared automated-controller changes** — runaway detection implemented inside `AutomatedAdjustmentController` evaluating only observations that follow an executed plan (so manual alignments and solve noise can't trip it), and correction limits modeled as an explicit capability supplied by the selected system rather than shared auto-scaling in `MoveCloser`. With manual, UPAS, and OAPA regression tests.

The test-file churn (`declinationSpreadArcsec: 0` noise) will be reverted and won't reappear in the series.

Thanks again for the detailed pass — the structural comments on the VM and the transport base in particular will make this considerably easier to maintain.

**Comment by michelebergo** [comment]
Closing in favor of the split series outlined above.
